### PR TITLE
[solidity] add int types

### PR DESCRIPTION
**Pull request by correcthorsebatterystaple**
Note:
This PR has branched off of (feature/add-solidity-uint-types) #754 which implements uint types and coercion for literal types to a specific uint type based on its context 

Changes:
- Add processing for int types in solidity
- Add regression tests for int types in solidity

### CODE REVIEW

**Review by lucasccordeiro — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
@rafaelsamenezes and @fbrausse: could you please review this PR?

**Comment by fbrausse** [comment]
@lucasccordeiro I'd like to resolve the comments for the first uint-PR first. It seems this PR duplicates many of the declarations of the previous one and actually includes all changes of some version of the previous PR as well.

**Comment by lucasccordeiro** [comment]
@correcthorsebatterystaple: could you please rebase this PR? Do you still have comments to implement?

File: src/solidity-frontend/solidity_convert.cpp [hunk 1]
```diff
@@ -467,7 +475,7 @@ bool solidity_convertert::get_block(
   get_start_location_from_stmt(block, location);
 
   SolidityGrammar::BlockT type = SolidityGrammar::get_block_t(block);
-  msg.debug(fmt::format(
+  msg.status(fmt::format(
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_convert.cpp]: nah... debug is fine; btw. with `--verbosity 7` you can enable debug messages.

File: src/solidity-frontend/solidity_convert.cpp [hunk 2]
```diff
@@ -819,6 +827,9 @@ bool solidity_convertert::get_expr(
     const nlohmann::json &literal = expr["typeDescriptions"];
     SolidityGrammar::ElementaryTypeNameT type_name =
       SolidityGrammar::get_elementary_type_name_t(literal);
+    msg.status(fmt::format(
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_convert.cpp]: `.debug()`, if at all; also below

File: src/solidity-frontend/solidity_convert.cpp [hunk 3]
```diff
@@ -1040,26 +1051,28 @@ bool solidity_convertert::get_binary_operator_expr(
...
   exprt &new_expr)
 {
-  // 1. get type
+  // TODO: Fix me! Currently just support prefix == true,e.g. pre-increment
+  assert(expr["prefix"]);
+
+  // 1. get UnaryOperation opcode
+  SolidityGrammar::ExpressionT opcode =
+    SolidityGrammar::get_unary_expr_operator_t(expr, expr["prefix"]);
+  msg.status(fmt::format(
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_convert.cpp]: debug

File: src/solidity-frontend/solidity_convert.cpp [hunk 4]
```diff
@@ -1357,7 +1375,7 @@ bool solidity_convertert::get_type_description(
   }
   default:
   {
-    msg.debug(fmt::format(
+    msg.status(fmt::format(
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_convert.cpp]: debug

File: src/solidity-frontend/solidity_convert.cpp [hunk 5]
```diff
@@ -1522,7 +1598,7 @@ bool solidity_convertert::get_elementary_type_name(
...
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_convert.cpp]: debug

File: src/solidity-frontend/solidity_grammar.cpp [hunk 6]
```diff
@@ -426,7 +491,35 @@ ExpressionT get_expression_t(const nlohmann::json &expr)
...
+  if(expr["operator"] == "-")
+  {
+    return UO_Minus;
+  }
+  assert(!((fmt::format(
+              "Got expression operator={}. Unsupported expression operator",
+              expr["operator"].get<std::string>()))
+             .c_str()));
+
+  return ExpressionTError; // make some old compilers happy
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_grammar.cpp]: I don't understand. You're asserting `false` just above and now `make some old compilers happy`? What compilers are meant here? Also new compilers will happily complain if this function doesn't return anything despite the `assert`. Asserts get compiled out on `Release` builds, old and new compilers will not know what to do then.

Either
- leave as it is without the comment, or
- print the error message and `abort()` - old and new compilers will be happy, or
- use `__builtin_unreachable` (least preferrable option)

File: src/solidity-frontend/solidity_grammar.cpp [hunk 7]
```diff
@@ -460,27 +553,9 @@ ExpressionT get_expr_operator_t(const nlohmann::json &expr, bool uo_pre)
...
-    {
-      return UO_PreDec;
-    }
-    else
-    {
-      assert(!"Unsupported - UO_PostDec");
-    }
-  }
-  else if(expr["operator"] == "++")
+  else if(expr["operator"] == "&&")
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_grammar.cpp:556]: That seems to be the same case as just below.

**Review by fbrausse — CHANGES_REQUESTED** [review]

File: src/solidity-frontend/solidity_grammar.cpp (hunk 7 above)
Comment (correcthorsebatterystaple) [reply src/solidity-frontend/solidity_grammar.cpp:556]: This is due to a bad merge, will remove.

File: src/solidity-frontend/solidity_grammar.cpp (hunk 6 above)
Comment (correcthorsebatterystaple) [reply src/solidity-frontend/solidity_grammar.cpp]: This is due to copy and paste from the source function which this has been extracted from. See below.

As the rest of the methods employ the assert I shall leave it in and remove the comment

Comment (fbrausse) [reply src/solidity-frontend/solidity_grammar.cpp]: Thanks!

File: src/solidity-frontend/solidity_convert_literals.cpp [hunk 8]
```diff
@@ -21,19 +21,12 @@ bool solidity_convertert::convert_integer_literal(
...
-      integer2binary(z_ext_value, bv_width(type)),
-      integer2string(z_ext_value),
-      type);
-  }
-  else
-  {
-    assert(!"Unimplemented - Got signed type. Add sigend bv conversion");
-  }
+  // extract the value: unsigned
+  BigInt z_ext_value = string2integer(the_value);
```
Comment (fbrausse) [inline src/solidity-frontend/solidity_convert_literals.cpp:25]: Watch out here: `string2integer()` only deals with strings matching `[0-9]*`, a leading `-` is not interpreted correctly. We should eventually fix this.

**Review by fbrausse — CHANGES_REQUESTED** [review]
Please revert all the msg.{debug => status} changes in this PR and also remove the `regression/int.patch` file.

File: src/solidity-frontend/solidity_convert_literals.cpp (hunk 8 above)
Comment (correcthorsebatterystaple) [reply src/solidity-frontend/solidity_convert_literals.cpp:25]: Seems like this doesn't cause an issue and it detects the underflow on the int types correctly

**Review by fbrausse — APPROVED** [review]
LGTM!

**Pull request merged**